blogposts/2020/code-search-turned-code-checker.md

```diff
@@ -0,0 +1,296 @@
+(false negatives). Each check may have more than one syntactic _variant_, so I
+tried to implement patterns for as many variants as I could find in tests. It's
+a neat exercise to develop patterns against the reference tests and discover
+which variants to cover, all in a self-contained search webapp. Here's an example
+where the query matches all the true hits in the test file, annotated with
+`// want strings.Contains ...`:
+
+
+<img src="https://storage.googleapis.com/sourcegraph-assets/about.sourcegraph.com/blog/2020/multiline-query-editor.png">
+
+[🔘 Example query to match known patterns in test files↗](https://sourcegraph.com/search/console?q=repo:%5Egithub%5C.com/dominikh/go-tools$%20%0Alang:go%0Afile:test%0A%0Astrings.IndexRune(...,%20...)%20%3E%20-1%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20%3E=%200%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20!=%20-1%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20==%20-1%0A%0Aor%0A%0Astrings.IndexRune(...,%20...)%20%3C%200%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20%3E%20-1%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20%3E=%200%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20!=%20-1%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20==%20-1%0A%0Aor%0A%0Astrings.IndexAny(...,%20...)%20%3C%200%0A%0Aor%0A%0Astrings.Index(...,%20...)%20%3E%20-1%0A%0Aor%0A%0Astrings.Index(...,%20...)%20%3E=%200%0A%0Aor%0A%0Astrings.Index(...,%20...)%20!=%20-1%0A%0Aor%0A%0Astrings.Index(...,%20...)%20==%20-1%0A%0Aor%0A%0Astrings.Index(...,%20...)%20%3C%200%0A)
+
+Using the test files isn't a guarantee that I've implemented all the checks or
+that it's entirely precise, but it adds a lot more confidence than inferring
+patterns only from documentation.
+
+### Results
+
+Without making any claims about completeness, I was able to implement
+at least one variant for 19 out of 34 checks checks that I feel confident
+about. I relied only on patterns in `staticcheck`'s test data to discover syntactic
+variants for checks, so I don't know if I covered all the patterns that
+`staticcheck` implements in its code.
+
+The majority of checks that I couldn't write required type information to
+implement correctly (13 of 34). Other checks I couldn't write required more
+complex syntax matching rules (8 of 34). This table roughly quantifies the
+expressive needs for implementing checks:
+
+| Description                      | # of checks |
+|----------------------------------|-------------|
+| Total                            | 34          |
+| Works (all variants)             | 11          |
+| Works (at least one variant)     | 19          |
+| Needs additional syntax matching | 8           |
+| Needs type info                  | 13          |
+
+<sup>Note that some checks require type info _and_ additional syntax matching.
+Also, one working variant for a check may not require type info, but another
+variant for the same check may. I.e., the values overlap and do not sum to
+Total.</sup>
```

Comment:
**nit**: since the previous and next paragraphs use "information" over "info", use it here as well for consistency. Re: may/may not, I find that starting with the affirmative flows better.

```suggestion
<sup>Note that some checks require type information _and_ additional syntax matching.
Additionally, one working variant for a check may require type information, while another
variant for the same check may not. I.e., the values overlap and do not sum to
Total.</sup>
```